pkg/provisioner/ironic/ironic.go

```diff
@@ -1152,6 +1152,47 @@ func (p *ironicProvisioner) GetFirmwareSettings(includeSchema bool) (settings me
 	return settings, schema, nil
 }
 
+// GetFirmwareComponents gets all available firmware components for a node and return a list.
+func (p *ironicProvisioner) GetFirmwareComponents() ([]metal3api.FirmwareComponentStatus, error) {
+	ironicNode, err := p.getNode()
+	if err != nil {
+		return nil, fmt.Errorf("could not get node to retrieve firmware components: %w", err)
+	}
+
+	// Get the components from Ironic via Gophercloud
+	var componentList []nodes.FirmwareComponent
+	var componentListErr error
+	if p.availableFeatures.HasFirmwareUpdates() {
+		componentList, componentListErr = nodes.ListFirmware(p.ctx, p.client, ironicNode.UUID).Extract()
+	} else {
+		return nil, fmt.Errorf("current ironic version does not support firmware updates")
+	}
+
+	if componentListErr != nil {
```

Comment:
This where I would try to get some granularity (see my comments on the controller). What do we return for nodes that don't have firmware support?

Reply:
Do you think it would make sense to check for example the driver in use and return a specific message?

Reply:
Well, we have firmwareInterface in BMCAccess, right? So we can call that and see if it returns "no-firmware".

Reply:
yeah, via AccessDetails , since it provides the FirmwareInterface https://github.com/metal3-io/baremetal-operator/blob/4cfd77682f48c528b9203642545db058915f38a0/pkg/hardwareutils/bmc/access.go#L77 , I will update to check the information

Reply:
I've updated to:
```
	if componentListErr != nil {

		bmcAccess, _ := p.bmcAccess()
		if bmcAccess.FirmwareInterface() == "no-firmware" {
			return nil, fmt.Errorf("node %s is using firmware interface %s: %w", ironicNode.UUID, bmcAccess.FirmwareInterface(), componentListErr)
		}

		return nil, fmt.Errorf("could not get firmware components for node %s: %w", ironicNode.UUID, componentListErr)
	}
```

Reply:
Looking at what I did when doing the implementation we return 404 when we don't find the information.. https://review.opendev.org/c/openstack/ironic/+/885276/7/api-ref/source/baremetal-api-v1-nodes-firmware.inc
I will re-install my bifrost setup tomorrow to test with ipmi to see what I would get

Reply:
@dtantsur I've tested with ipmi, the response from the API is an empty list

```
X-Openstack-Ironic-Api-Minimum-Version: 1.1
X-Openstack-Ironic-Api-Maximum-Version: 1.90
X-Openstack-Ironic-Api-Version: 1.86
Content-Length: 16
Content-Type: application/json
Openstack-Request-Id: req-fa19ac25-da8f-449c-a098-2b01fb3db61d
Date: Fri, 15 Mar 2024 15:47:44 GMT

{"firmware": []}
```
This is the output of a machine with redfish
```
X-Openstack-Ironic-Api-Minimum-Version: 1.1
X-Openstack-Ironic-Api-Maximum-Version: 1.89
X-Openstack-Ironic-Api-Version: 1.86
Content-Length: 435
Content-Type: application/json
Openstack-Request-Id: req-10ddbf4a-7ca4-4a22-8de9-7b65019b3bdc
Date: Fri, 15 Mar 2024 15:49:01 GMT

{"firmware": [{"created_at": "2023-10-03T18:30:00+00:00", "updated_at": null, "component": "bios", "initial_version": "U30 v2.36 (07/16/2020)", "current_version": "U30 v2.36 (07/16/2020)", "last_version_flashed": null}, {"created_at": "2023-10-03T18:30:00+00:00", "updated_at": "2023-10-03T18:45:54+00:00", "component": "bmc", "initial_version": "iLO 5 v2.78", "current_version": "iLO 5 v2.81", "last_version_flashed": "iLO 5 v2.81"}]}
```

In case the node doesn't exist:
```
X-Openstack-Ironic-Api-Minimum-Version: 1.1
X-Openstack-Ironic-Api-Maximum-Version: 1.90
X-Openstack-Ironic-Api-Version: 1.86
Openstack-Request-Id: req-22c9105c-e457-40ab-8cdd-6c657b0005f1
Content-Type: application/json
Content-Length: 128
Date: Fri, 15 Mar 2024 15:49:53 GMT

{"error_message": "{\"faultcode\": \"Client\", \"faultstring\": \"Field fi2mware is not a valid field.\", \"debuginfo\": null}"}
```

Reply:
Sigh. I don't like the fact that Ironic does not return an error, but the ship has sailed.

Reply:
we probably only validate if has a "valid" interface in use